compiler/rustc_resolve/src/late.rs

```diff
@@ -1495,17 +1528,17 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
 
             match rib.kind {
                 LifetimeRibKind::Item => break,
-                LifetimeRibKind::ConstGeneric => {
-                    self.emit_non_static_lt_in_const_generic_error(lifetime);
+                LifetimeRibKind::ConstParamTy => {
+                    self.emit_non_static_lt_in_const_param_ty_error(lifetime);
                     self.record_lifetime_res(
                         lifetime.id,
                         LifetimeRes::Error,
                         LifetimeElisionCandidate::Ignore,
                     );
                     return;
                 }
-                LifetimeRibKind::AnonConst => {
-                    self.maybe_emit_forbidden_non_static_lifetime_error(lifetime);
+                LifetimeRibKind::ConcreteAnonConst(cause) => {
```

Comment:
> Every lifetime you write that isnt '_ or 'static gets resolved to an error without an error actually having been emitted, this goes for when you do actually have a lifetime in scope (i.e. fn foo<'a>() -> [(); { let _: &'a (); 10 }] {}) or when you are writing lifetimes that haven't been declared (i.e. fn foo() -> [(); { let _: &'a (); 10 }] {}). 

The bug is here. We need the feature check for `generic_const_exprs` be done on this whole block, instead of only doing it in `emit_forbidden_non_static_lifetime_error`. This way, lifetime resolution will continue walking ribs, instead of returning a `LifetimeRes::Error` straight.

Reply:
yeah, this PR accomplishes that by just not adding the `ConcreteAnonConst` rib if `generic_const_exprs` is enabled, so the only time we encounter `LifetimeRibKind::ConcreteAnonConst` should be when we need to error